Sources/MongoSwift/MongoError.swift

```diff
@@ -2,277 +2,287 @@ import CLibMongoC
 
-/// An error thrown when the driver encounters a internal error not caused by the user. This is usually indicative of
-/// a bug in the driver or system related failure (e.g. memory allocation failure).
-public struct InternalError: RuntimeError {
-    internal let message: String
+    /// An error thrown when encountering a connection or socket related error.
+    /// May contain labels providing additional information on the nature of the error.
+    public struct ConnectionError: MongoRuntimeError, MongoLabeledError {
+        public let message: String
 
-    public var errorDescription: String? { self.message }
-}
+        public let errorLabels: [String]?
 
-/// An error thrown when encountering a connection or socket related error.
-/// May contain labels providing additional information on the nature of the error.
-public struct ConnectionError: RuntimeError, LabeledError {
-    public let message: String
+        public var errorDescription: String? { self.message }
+    }
 
-    public let errorLabels: [String]?
+    /// An error thrown when encountering an authentication related error (e.g. invalid credentials).
+    public struct AuthenticationError: MongoRuntimeError {
+        internal let message: String
 
-    public var errorDescription: String? { self.message }
-}
+        public var errorDescription: String? { self.message }
+    }
 
-/// An error thrown when encountering an authentication related error (e.g. invalid credentials).
-public struct AuthenticationError: RuntimeError {
-    internal let message: String
+    /// An error thrown when trying to use a feature that the deployment does not support.
+    public struct CompatibilityError: MongoRuntimeError {
+        internal let message: String
 
-    public var errorDescription: String? { self.message }
-}
-
-/// An error thrown when trying to use a feature that the deployment does not support.
-public struct CompatibilityError: RuntimeError {
-    internal let message: String
+        public var errorDescription: String? { self.message }
+    }
 
-    public var errorDescription: String? { self.message }
-}
+    /// An error that occured when trying to select a server (e.g. a timeout, or no server matched read preference).
+    ///
+    /// - SeeAlso: https://docs.mongodb.com/manual/core/read-preference-mechanics/
+    public struct ServerSelectionError: MongoRuntimeError {
+        internal let message: String
 
-/// An error that occured when trying to select a server (e.g. a timeout, or no server matched read preference).
-///
-/// - SeeAlso: https://docs.mongodb.com/manual/core/read-preference-mechanics/
-public struct ServerSelectionError: RuntimeError {
-    internal let message: String
+        public var errorDescription: String? { self.message }
+    }
 
-    public var errorDescription: String? { self.message }
-}
+    /// A struct to represent a single write error not resulting from an executed write operation.
+    public struct WriteFailure: Codable {
```

Comment:
these `-Failure` types aren't errors themselves, but they are contained within different error types. I wasn't sure where to put them so I just left them under `MongoError`.